core/src/metrics/registry.rs

```diff
@@ -197,6 +202,21 @@ impl MetricsRegistryTrait for MetricsRegistry {
         Ok(counter)
     }
 
+    fn global_counter(&self, name: String) -> Result<Counter, PrometheusError> {
+        let maybe_counter = self.global_counters.read().unwrap().get(&name).cloned();
+        if let Some(counter) = maybe_counter {
+            Ok(counter.clone())
+        } else {
+            let help = "global counter".to_owned();
+            let counter = *self.new_counter(name.clone(), help, HashMap::new())?;
+            self.global_counters
+                .write()
+                .unwrap()
+                .insert(name.clone(), counter.clone());
+            Ok(counter)
+        }
+    }
+
```

Comment:
I wonder if this could be made more atomic with e.g. `.entry(&name).or_insert_with()` but with the error handling and the `RwLock` around the hash map I don't see an obvious way.

Reply:
Yhea that's the first thing I tried but doesn't work for the reasons you mention.